### ISSUE TITLE
[Question] "Before" result, reused for multiple scenarios

### ISSUE THREAD

**Issue by Dynom** [Codeception/CodeceptJS#659]
### TL;DR
I have a setup where I want to run many Scenario's against something (a container of sorts) that is created. This creation is quite expensive (time wise) and I would like to create just one for the entire suite to run against.

### Expensive setup
I would like to create a container each time my test suite starts (so that each run is completely isolated, concurrency of having multiple containers isn't a problem). Since the creation of these containers are expensive, I want to re-use this container in many Scenarios during one run of the test-suite. Each Scenario won't influence the container in a way that could influence a different Scenario.

These containers take many complex steps to create, these are covered in tests. However there are many scenarios that can be run once such container exists. For this last part I would not like to re-create a container for every Scenario.

### To create these containers I need
- Access to a browser, I need certain details to create a container (E.g. a pre-configured URL and an access token)
- To perform (several) requests to a web service (`I.sendPostRequest()` performs adequately), each response contains information I need for the next.
- Capture the final output, as it contains (at the very least) a URL on which to base scenarios on (`I.amOnPage(container.rootURL);`)

### Things I've tried / considered
- `BeforeSuite` seems like an semantically acceptable candidate, however I don't have access to a browser. 
- `Step objects` I can't get to work. Promise chaining doesn't work with the `I.*` methods and I need to pass in values obtained from `I.*` functionality in a Scenario and I need the result of the actions from the step object.
- `Factories` Creating a factory for this would probably work, but seems perhaps like overkill and a non-trivial effort.

So how can I accomplish this complex setup which would be idiomatic CodeceptJS?

**Comment by APshenkin** [comment]
@Dynom 
Hi!

1. You will have access to browser in `BeforeSuite`, if you will use option `restart:false`
2. I can not quite understand, what does you mean, when you use term `container` (is it Docker container?) and also I don't understand why you need access to browser? Why you can't use usual requests? BTW if you need to start docker containers or any other environment, I suggest to use bootstrap file for it

**Comment by Dynom** [comment]
Hi @APshenkin,

Thanks for your reply! I'll try and see if can make it work with suggestion 1.

**To answer point 2**
The term `container` is more an abstract term for the problem I'm describing. The crux is that I need to access a webpage to collect certain details that are only temporarily available. The situation I'm testing is part of an integration test. The user-journey I'm testing, requires the tests to obtain short-lived details from a webpage:

* A unique URL, which serves as the starting point for the tests (`I.amOnPage(uniqueURL);`)
* An access token (`I.haveRequestHeaders({..});`)

These details I need to communicate with a web-service to prepare the data for the test. Once the data is prepared, it can be used for many tests. The preparation of this data is, however, costly.

There are several ways of solving this:

1. Pull the work of preparing outside of CodeceptJS and make it part of the bootstrapping. I could have a Makefile that handles the load. The downside is that I need to maintain certain knowledge in two places, something I'd like to avoid.
2. Do the preparation inside CodeceptJS. The downside for this approach is that it doesn't (seem to) fall in the responsibility domain that CodeceptJS addresses.
3. Pull the work outside of CodeceptJS in a more extreme way. Prepare an environment with completely prepared data, including a provisioned database where all details are pre-configured and known in advance. This is something we're working towards but it requires a large investment and lot of effort to setup.

To illustrate it a bit further, this is a scenario that works, but I'd like to refactor it so that it's much more reusable:
```js
Scenario('Change setting Foo', function *(I) {
    // ---
    // @todo, refactor this to a Feature-wide re-usable resource.

    // Cloning the request template
    let r1 = Object.assign({}, requestA);
    let r2 = Object.assign({}, requestB);

    I.amOnPage(createContainerURL);

    const authenticationToken = yield I.grabAttributeFrom('//body', 'data-token');
    const brandName = yield I.grabAttributeFrom('//body', 'data-brand');

    I.haveRequestHeaders({
        'token': authenticationToken,
    });

    // Preparing first request
    r1.brand = brandName;

    // Sending first request, creating resource
    const result1 = yield I.sendPostRequest(containerAPIURL, JSON.stringify(r1));

    // Preparing second request
    r2.brand = brandName;
    r2.settings = { title: initialTitle };

    // Sending second request, updating resource
    const result2 = yield I.sendPostRequest(containerAPIURL, JSON.stringify(r2));

    const result = result2.body.result;

    // Event created, starting the test
    // ---

    I.amOnPage(result.rootURL);

    // Verify the title is set to initialTitle
    I.seeElement(`//*[@id="main_details"]//h1[@class="title"][contains(., "${initialTitle}")]`);

    // Change the title
    // ..

    // Verify the title was changed
    // ..
```

**Comment by reubenmiller** [comment]
Closing ticket due to inactivity (after a provided solution)

**Comment by Dynom** [comment]
In case anyone ends up on the question and is curious about the answer, we diverted to using PageObjects and plain-old-JavaScript to write the functionality. We can now do:

```JavaScript
Scenario("Testing functionality Foo", async (I, eventFactory) => {
  const b = eventFactory.getBuilder();
  b.setTitle('foo');
  const eventDetails = await eventFactory.create(b);

  I.amOnPage(eventDetails.getEventUrl());
  I.seeElement(`//h1[@class="title"][contains(., "${eventDetails.event.title}")]`);

  // ..
});
```
